runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1318,240 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+	 * 1 implies that eden should move all the way towards edenChange.
+	 */
+	intptr_t edenChange = calculateRecommendedEdenChangeForExpandedHeap(env);
+	intptr_t targetEdenChange = (intptr_t)(edenChange * edenChangeSpeed);
+	uintptr_t targetEdenBytes = currentIdealEdenBytes + targetEdenChange;
+	uintptr_t targetEdenRegions = targetEdenBytes / _regionManager->getRegionSize();
+
+	return (intptr_t)targetEdenRegions - (intptr_t)currentIdealEdenRegions;
+}
+
+void
+MM_SchedulingDelegate::checkEdenSizeAfterPgc(MM_EnvironmentVLHGC *env, bool globalSweepHappened)
+{
+	if (_currentlyPerformingGMP && !globalSweepHappened) {
+		/* 
+		 * Don't change eden size while GMP cycle is running - 
+		 * Unless a global sweep just happened, in which case we allow eden to change on first PGC after GMP 
+		 */
+		return;
+	}
+
+	double ratioOfHeapExpanded = 0.0;
+	uintptr_t currentHeapSize = _regionManager->getRegionSize() * _numberOfHeapRegions;
+	uintptr_t minimumHeap = OMR_MIN(_extensions->initialMemorySize, currentHeapSize);
+	uintptr_t maximumHeap = _extensions->softMx == 0 ? _extensions->memoryMax : _extensions->softMx;
+
+	if ((maximumHeap == currentHeapSize) || (maximumHeap == minimumHeap)) {
+		ratioOfHeapExpanded = 1.0;
+	} else {
+		uintptr_t heapBytesOverMinimum = currentHeapSize - minimumHeap;
+		uintptr_t maximumHeapVariation = maximumHeap - minimumHeap;
+		ratioOfHeapExpanded = ((double)heapBytesOverMinimum) / ((double)maximumHeapVariation);
+	}
+
+	double heapFullyExpandedThreshold = 0.9;
+	double heapPercentExpandedAboveThreshold = 0.0;
+	if (heapFullyExpandedThreshold < ratioOfHeapExpanded) {
+		heapPercentExpandedAboveThreshold = ratioOfHeapExpanded - heapFullyExpandedThreshold;
+	}
+
+	intptr_t heapNotFullyExpandedRecommendation = 0;
+	intptr_t heapFullyExpandedRecommendation = 0;
+
+	if (0.0 == heapPercentExpandedAboveThreshold) {
+		heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+	} else if (0.0 < heapPercentExpandedAboveThreshold) {
+		/** 
+		 * If heap is almost fully expanded, start to consider the recommendation given from moveTowardRecommendedEdenForExpandedHeap().
+		 */
+		if (globalSweepHappened) {
+			heapFullyExpandedRecommendation = moveTowardRecommendedEdenForExpandedHeap(env, 0.5);
+			heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+			resetPgcTimeStatistics(env);
+		} else if (0 == _pgcCountSinceGMPEnd % consecutivePGCToChangeEden){
+			/** 
+	 		 * Every consecutivePGCToChangeEden number of PGC's, check to see if eden size should change. 
+			 * This allows PGC to do some defragmentation work to finish, and for some statistics to stabilize
+ 			 */
+			heapFullyExpandedRecommendation = moveTowardRecommendedEdenForExpandedHeap(env, 0.25);
+			heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+		}
+	}
+
+	/** 
+	 * Once ratioOfHeapExpanded is over heapFullyExpandedThreshold, start to blend the recommendations provided by 
```

Comment:
explain not only what we do it by why. something like: to ensure smooth transition of potentially very different recommended eden sizes for fully expanded case and non-expanded heap case when the cases frequently change from one to the other

Reply:
Comment has been updated. See just a few lines below